src/chains/algorand_1/algoChainState.ts

```diff
@@ -183,4 +180,29 @@ export class AlgorandChainState {
     this.assertIsConnected()
     return this._algoClient
   }
+
+  // TODO: sort based on health info
+  /**  * Choose the best Chain endpoint based on health and response time */
+  private selectEndpoint(): { url: URL; endpoint: AlgorandChainEndpoint } {
+    // Just choose the first endpoint for now
+    const endpoint = this.endpoints[0]
+    const url = endpoint?.url?.href
+    return { url: new URL(trimTrailingChars(url, '/')), endpoint }
+  }
+
+  /** returns the required header from the array of objects. For ex: headers: [{'X-API-Key': '...'}]  */
+  private getHeader(key: string) {
+    const { headers } = this._activeEndpoint?.options
+    const header = headers.find((val: {}) => Object.keys(val).includes(key))
+    return header
+  }
+
+  /** returns the 'X-API-Key' header required to call algorand chain endpoint */
+  private getAlgorandConnectionHeader(): AlgorandHeader {
```

Comment:
This function returns the token header so the name is misleading. It also does an unexpected thing. You don't expect a get to validate/throw.
I'd have two functions. getTokenFromEndpointHeader() and assertEndpointHasTokenHeader()
Put assertEndpointHasTokenHeader() in connect() right after the call to selectEndpoint
If getTokenFromEndpointHeader() can't find a token header, just return null

Reply:
done